README.md
## Angle utils

A small and compact utility library for dealing with all kinds of angles. This library supports radians, degrees, turns and gradians.

### Angle

```dart
var angle = Angle.degrees(90);
angle.radians;  // pi/2
angle.degrees;  // 90
angle.gradians; // 100
angle.turns;    // 1/4

var full = Angle.full();
full.radians;   // 2*pi
full.degrees;   // 360
full.gradians;  // 400
full.turns;     // 1

var diff = full - angle;
diff.degrees; // 270

var other = Angle.degrees(-10);
other.normalized.degrees; // 350
```

### AngleRange

This library also provides an AngleRange type:

```dart
var range = AngleRange(start: Angle.degrees(80), end: Angle.degrees(100));
range.start;    // Angle.degrees(80)
range.end;      // Angle.degrees(100)
range.mid;      // Angle.degrees(90)
range.sweep;    // Angle.degrees(20)

range.includes(Angle.degrees(90)); // true
range.includes(Angle.degrees(70)); // false
```

CHANGELOG.md
## 0.0.4
- Added ratio and getClosest function

## 0.0.3
- More documentation

## 0.0.2
- Added more documentation and fixed formatting

## 0.0.1
- First release


test/angle_test.dart
import 'dart:math' as math;

import 'package:angle_utils/angle_utils.dart';
import 'package:test/test.dart';

void main() {
  group('Creating and converting angle', () {
    var angle0 = Angle.degrees(0);
    var angle90 = Angle.radians(math.pi / 2);
    var angle180 = Angle.degrees(180);
    var angle270 = Angle.gradians(300);
    var angle360 = Angle.turns(1);

    test('.degrees(0) should equal 0 radians, 0 gradians and 0 turns', () {
      expect(angle0.radians, 0);
      expect(angle0.gradians, 0);
      expect(angle0.turns, 0);
    });

    test('.radians(pi/2) should equal 90 degrees, 100 gradians and 0.25 turns',
        () {
      expect(angle90.degrees, 90.0);
      expect(angle90.gradians, 100);
      expect(angle90.turns, 0.25);
    });

    test('.degrees(180) should equal pi radians, 200 gradians and 0.5 turns',
        () {
      expect(angle180.radians, math.pi);
      expect(angle180.gradians, 200);
      expect(angle180.turns, 0.5);
    });

    test(
        '.gradians(300) should equal 3*pi/2 radians, 270 degrees and 0.75 turns',
        () {
      expect(angle270.radians, 3 * math.pi / 2);
      expect(angle270.degrees, 270);
      expect(angle270.turns, 0.75);
    });

    test('.turns(1) should equal 2*pi radians, 400 gradians and 360 degrees',
        () {
      expect(angle360.radians, 2 * math.pi);
      expect(angle360.gradians, 400);
      expect(angle360.degrees, 360);
    });
  });

  test('.normalized', () {
    expect(Angle.a0().normalized, Angle.a0());
    expect(Angle.a90().normalized, Angle.a90());
    expect(Angle.degrees(450).normalized, Angle.a90());
    expect(Angle.degrees(720).normalized, Angle.a0());
    expect(Angle.degrees(-10).normalized, Angle.degrees(350));
    expect(Angle.degrees(-380).normalized, Angle.degrees(340));
  });

  test('.complementary', () {
    expect(Angle.a0().complementary, Angle.a90());
    expect(Angle.degrees(10).complementary, Angle.degrees(80));
    expect(Angle.degrees(80).complementary, Angle.degrees(10));
    expect(Angle.degrees(25.4).complementary, Angle.degrees(64.6));
    expect(Angle.degrees(100).complementary, Angle.degrees(-10));
  });

  test('.supplementary', () {
    expect(Angle.a0().supplementary, Angle.a180());
    expect(Angle.degrees(80).supplementary, Angle.degrees(100));
    expect(Angle.degrees(170).supplementary, Angle.degrees(10));
    expect(Angle.degrees(90.5).supplementary, Angle.degrees(89.5));
    expect(Angle.degrees(200).supplementary, Angle.degrees(-20));
  });

  test('.distance', () {
    expect(Angle.getMinimalDistance(Angle.a0(), Angle.a90()), Angle.a90());
    expect(Angle.getMinimalDistance(Angle.degrees(45), Angle.degrees(75)),
        Angle.degrees(30));
    expect(Angle.getMinimalDistance(Angle.degrees(270), Angle.degrees(180)),
        Angle.a90());
    expect(Angle.getMinimalDistance(Angle.degrees(350), Angle.degrees(10)),
        Angle.degrees(20));
    expect(Angle.getMinimalDistance(Angle.degrees(-90), Angle.degrees(45)),
        Angle.degrees(135));
  });

  test('.abs', () {
    expect(Angle.a90().abs(), Angle.a90());
    expect(Angle.degrees(-50).abs(), Angle.degrees(50));
  });

  test('adding angles', () {
    expect(Angle.a90() + Angle.a90(), Angle.a180());
    expect(Angle.degrees(350) + Angle.degrees(20), Angle.degrees(370));
    expect(Angle.degrees(-30) + Angle.degrees(10), Angle.degrees(-20));
    expect(Angle.degrees(-30) + Angle.degrees(-20), Angle.degrees(-50));
  });

  test('substracting angles', () {
    expect(Angle.a90() - Angle.a90(), Angle.a0());
    expect(Angle.a180() - Angle.degrees(45), Angle.degrees(135));
    expect(Angle.degrees(10) - Angle.degrees(45), Angle.degrees(-35));
    expect(Angle.degrees(-30) - Angle.degrees(20), Angle.degrees(-50));
  });

  test('multiplying angles', () {
    expect(Angle.a90() * 3, Angle.a270());
    expect(Angle.degrees(10) * 5, Angle.degrees(50));
    expect(Angle.degrees(-30) * 4, Angle.degrees(-120));
    expect(Angle.degrees(-10) * -5, Angle.degrees(50));
  });

  test('dividing angles', () {
    expect(Angle.a90() / 2, Angle.degrees(45));
    expect(Angle.a0() / 2, Angle.a0());
    expect(Angle.degrees(-10) / 5, Angle.degrees(-2));
    expect(Angle.degrees(-50) / -2, Angle.degrees(25));
  });

  test('ratio', () {
    expect(Angle.a90().ratio(Angle.a180()), 0.5);
    expect(Angle.a180().ratio(Angle.a180()), 1);
    expect(Angle.a180().ratio(Angle.a90()), 2);
    expect(Angle.a360().ratio(Angle.a90()), 4);
  });

  test('getClosest', () {
    var angles = [
      Angle.degrees(45),
      Angle.degrees(90),
      Angle.degrees(135),
      Angle.degrees(245),
      Angle.degrees(350),
    ];
    expect(Angle.degrees(10).getClosest(angles), Angle.degrees(350));
    expect(Angle.degrees(30).getClosest(angles), Angle.degrees(45));
    expect(Angle.degrees(60).getClosest(angles), Angle.degrees(45));
    expect(Angle.degrees(100).getClosest(angles), Angle.degrees(90));
    expect(Angle.degrees(135).getClosest(angles), Angle.degrees(135));
    expect(Angle.degrees(200).getClosest(angles), Angle.degrees(245));
    expect(Angle.degrees(360).getClosest(angles), Angle.degrees(350));
    expect(Angle.degrees(190).getClosest(angles), Angle.degrees(135));

    expect(() => Angle.degrees(10).getClosest([]), throwsArgumentError);
  });
}


test/angle_range_test.dart
import 'package:angle_utils/angle_utils.dart';
import 'package:test/test.dart';

void main() {
  var range1 = AngleRange(start: Angle.a0(), end: Angle.a90());
  var range2 = AngleRange(start: Angle.a180(), end: Angle.degrees(280));
  var range3 = AngleRange(start: Angle.a270(), end: Angle.degrees(10));
  var range4 = AngleRange(start: Angle.degrees(-40), end: Angle.degrees(-60));

  test('start and end', () {
    expect(range1.start, Angle.a0());
    expect(range1.end, Angle.a90());

    expect(range2.start, Angle.a180());
    expect(range2.end, Angle.degrees(280));

    expect(range3.start, Angle.a270());
    expect(range3.end, Angle.degrees(10));

    expect(range4.start, Angle.degrees(-40));
    expect(range4.end, Angle.degrees(-60));
  });

  test('sweep', () {
    expect(range1.sweep, Angle.a90());
    expect(range2.sweep, Angle.degrees(100));
    expect(range3.sweep, Angle.degrees(100));
    expect(range4.sweep, Angle.degrees(340));
  });

  test('mid', () {
    expect(range1.mid, Angle.degrees(45));
    expect(range2.mid, Angle.degrees(230));
    expect(range3.mid, Angle.degrees(320));
    expect(range4.mid, Angle.degrees(130));
  });

  test('fromSweep', () {
    var range = AngleRange.fromSweep(Angle.degrees(100), Angle.degrees(50));

    expect(range.start, Angle.degrees(100));
    expect(range.end, Angle.degrees(150));
    expect(range.mid, Angle.degrees(125));
    expect(range.sweep, Angle.degrees(50));
  });

  test('fromDelta', () {
    var range = AngleRange.fromDelta(Angle.degrees(45), Angle.degrees(10));

    expect(range.start, Angle.degrees(35));
    expect(range.end, Angle.degrees(55));
    expect(range.mid, Angle.degrees(45));
    expect(range.sweep, Angle.degrees(20));
  });

  test('normalized', () {
    var range = AngleRange(start: Angle.degrees(-10), end: Angle.degrees(-5))
        .normalized;

    expect(range.start, Angle.degrees(350));
    expect(range.end, Angle.degrees(355));
    expect(range.mid, Angle.degrees(352.5));
    expect(range.sweep, Angle.degrees(5));

    expect(range, AngleRange.normalized(Angle.degrees(-10), Angle.degrees(-5)));
  });

  test('includesNormalized', () {
    var range = AngleRange(start: Angle.degrees(370), end: Angle.degrees(390));

    expect(range.includesNormalized(Angle.degrees(360)), false);
    expect(range.includesNormalized(Angle.degrees(400)), false);
    expect(range.includesNormalized(Angle.degrees(390)), true);
    expect(range.includesNormalized(Angle.degrees(370)), true);
    expect(range.includesNormalized(Angle.degrees(380)), true);
    expect(range.includesNormalized(Angle.degrees(740)), true);
    expect(range.includesNormalized(Angle.degrees(15)), true);
  });

  test('includes', () {
    var range = AngleRange(start: Angle.degrees(370), end: Angle.degrees(390));

    expect(range.includes(Angle.degrees(360)), false);
    expect(range.includes(Angle.degrees(400)), false);
    expect(range.includes(Angle.degrees(390)), true);
    expect(range.includes(Angle.degrees(370)), true);
    expect(range.includes(Angle.degrees(380)), true);
    expect(range.includes(Angle.degrees(740)), false);
    expect(range.includes(Angle.degrees(15)), false);
  });

  test('errors', () {
    expect(() => AngleRange.fromDelta(Angle.degrees(90), Angle.degrees(-10)),
        throwsArgumentError);
    expect(() => AngleRange(start: Angle.degrees(0), end: Angle.degrees(370)),
        throwsArgumentError);
  });
}


test/angle_utils_test.dart
import 'dart:math' as math;

import 'package:angle_utils/angle_utils.dart';
import 'package:test/test.dart';

void main() {
  group('Angle', () {
    var angle0 = Angle.degrees(0);
    var angle90 = Angle.radians(math.pi/2);
    var angle180 = Angle.degrees(180);

    test('.degrees(0) should equal 0 radians', () {
      expect(angle0.radians, 0);
    });

    test('.radians(pi/2) should equal 90 degrees', () {
      expect(angle90.degrees, 90.0);
    });

    test('.degrees(180) should equal pi radians', () {
      expect(angle180.radians, math.pi);
    });
  });
}


lib/src/angle_range.dart
import 'angle.dart';

/// Represents an angle range.
class AngleRange {
  late final Angle _start;
  late final Angle _sweep;
  late final Angle _end;
  late final Angle _mid;

  Angle get start => _start;
  Angle get sweep => _sweep;
  Angle get end => _end;
  Angle get mid => _mid;

  AngleRange get normalized =>
      AngleRange(start: _start.normalized, end: _end.normalized);

  /// Constructs an angle range from the given [start] and [end] angle
  /// An angle range can NOT be over 360 degrees
  AngleRange({required Angle start, required Angle end}) {
    _start = start;
    _end = end;
    if ((_end - _start).abs() > Angle.full()) {
      throw ArgumentError('Angle range can not be greater than 360 degrees');
    }
    _sweep = (_end - _start).normalized;
    _mid = _start + _sweep / 2;
  }

  /// Constructs and angle range from a [start] and a [sweep] angle
  factory AngleRange.fromSweep(Angle start, Angle sweep) {
    return AngleRange(start: start, end: start + sweep);
  }

  /// Constructs an angle range from the given [mid] angle by adding and substracting [delta].
  factory AngleRange.fromDelta(Angle mid, Angle delta) {
    if (delta.isNegative) {
      throw ArgumentError('Delta angle must be positiv');
    }
    return AngleRange(start: mid - delta, end: mid + delta);
  }

  /// Constructs an angle range where [start] and [end] are normalized
  factory AngleRange.normalized(Angle start, Angle end) {
    return AngleRange(start: start.normalized, end: end.normalized);
  }

  /// Checks whether the given [angle] is in this normalized angle range
  /// Note: [start] can be greater than [end], but then the range spans over angle zero
  bool includesNormalized(Angle angle) {
    var normalizedRange = normalized;
    var normalizedAngle = angle.normalized;
    if (normalizedRange.start < normalizedRange.end) {
      return normalizedRange.start <= normalizedAngle &&
          normalizedAngle <= normalizedRange.end;
    } else if (normalizedRange.start > normalizedRange.end) {
      return normalizedRange.start <= normalizedAngle ||
          normalizedAngle <= normalizedRange.end;
    }
    return normalizedAngle == normalizedRange.start;
  }

  /// Checks whether the given [angle] is in this angle range
  bool includes(Angle angle) {
    return _start <= angle && angle <= _end;
  }

  @override
  bool operator ==(Object other) {
    return (other is AngleRange) &&
        other._start == _start &&
        other._end == _end;
  }

  @override
  int get hashCode => Object.hash(_start, _end);

  @override
  String toString() => '$_start to $_end';
}


lib/src/angle.dart
import 'dart:core';
import 'dart:math' as math;

import 'num_extension.dart';

/// Represents an angle.
/// Internally the angle is stored in radians.
class Angle implements Comparable<Angle> {
  final double _radians;

  bool get isZero => _radians == 0;

  bool get isPositive => _radians >= 0;
  bool get isNegative => _radians < 0;

  /// Returns the normalized angle:
  /// 45° -> 45°
  /// 370° -> 10°
  /// -90° -> 270°
  Angle get normalized {
    var remainder = _radians.remainder(2 * math.pi);
    if (isPositive) {
      return Angle.radians(remainder);
    } else {
      return Angle.full() + Angle.radians(remainder);
    }
  }

  /// Returns the complementary angle of this
  Angle get complementary => Angle.a90() - this;

  /// Returns the supplementary angle of this
  Angle get supplementary => Angle.a180() - this;

  double get radians => _radians;
  double get degrees => Angle.radiansToDegrees(_radians);
  double get turns => Angle.radiansToTurns(_radians);
  double get gradians => Angle.radiansToGradians(_radians);

  Angle._({required double radians}) : _radians = radians;

  /// Constructs an angle from degrees
  Angle.degrees(final double degrees)
      : _radians = Angle.degreesToRadians(degrees);

  /// Constructs an angle from turns
  Angle.turns(final double turns) : _radians = Angle.turnsToRadians(turns);

  /// Constructs an angle from gradians
  Angle.gradians(final double gradians)
      : _radians = Angle.gradiansToRadians(gradians);

  /// Constructs an angle from radians
  Angle.radians(final double radians) : _radians = radians;

  Angle.atan2(num a, num b) : _radians = math.atan2(a, b);

  /// Constructs an angle with 0 degrees
  factory Angle.zero() => Angle._(radians: 0);

  /// Constructs an angle with 180 degrees
  factory Angle.half() => Angle._(radians: math.pi);

  /// Constructs an angle with 360 degrees
  factory Angle.full() => Angle._(radians: 2 * math.pi);

  /// Constructs an angle with 0 degrees
  factory Angle.a0() => Angle.zero();

  /// Constructs an angle with 90 degrees
  factory Angle.a90() => Angle._(radians: math.pi / 2);

  /// Constructs an angle with 180 degrees
  factory Angle.a180() => Angle.half();

  /// Constructs an angle with 270 degrees
  factory Angle.a270() => Angle._(radians: 3 * math.pi / 2);

  /// Constructs an angle with 360 degrees
  factory Angle.a360() => Angle.full();

  /// Returns the distance between [a] and [b].
  /// The returned angle can never be over 180 degrees.
  static Angle getMinimalDistance(Angle a, Angle b) {
    // min{|α−β|, 360° −|α−β|}
    var diff = (a.normalized.radians - b.normalized.radians).abs();
    return Angle._(radians: math.min(diff, 2 * math.pi - diff));
  }

  /// Gets the closest angle out of [angles] to this.
  /// [angles] must contain at least one angle
  /// IMPORTANT: All angles are normalized!
  Angle getClosest(final List<Angle> angles) {
    if (angles.isEmpty) {
      throw ArgumentError('Angle list must contain at least one angle');
    }
    return angles.fold(angles[0], (previousAngle, currentAngle) {
      if (Angle.getMinimalDistance(this, currentAngle) <
          Angle.getMinimalDistance(this, previousAngle)) {
        return currentAngle;
      }
      return previousAngle;
    });
  }

  Angle abs() => Angle._(radians: _radians.abs());

  double sin() => math.sin(_radians);
  double cos() => math.cos(_radians);
  double tan() => math.tan(_radians);

  Angle operator -(Angle other) => Angle._(radians: _radians - other.radians);
  Angle operator -() => Angle._(radians: -_radians);
  Angle operator +(Angle other) => Angle._(radians: _radians + other.radians);
  Angle operator /(num divisor) => Angle._(radians: _radians / divisor);
  Angle operator *(dynamic factor) => Angle._(radians: _radians * factor);

  double ratio(final Angle other) => radians / other.radians;

  bool operator >(Angle other) => _radians > other.radians;
  bool operator <(Angle other) => _radians < other.radians;
  bool operator >=(Angle other) => _radians >= other.radians;
  bool operator <=(Angle other) => _radians <= other.radians;

  @override
  bool operator ==(Object other) {
    return (other is Angle) &&
        _radians.toPrecision(10) == other._radians.toPrecision(10);
  }

  @override
  int compareTo(Angle other) {
    if (this > other) {
      return 1;
    } else if (this < other) {
      return -1;
    }
    return 0;
  }

  @override
  int get hashCode => _radians.hashCode;

  @override
  String toString() => '${degrees.toStringAsFixed(2)}°';

  static double radiansToDegrees(final double radians) =>
      radians / math.pi * 180;
  static double radiansToTurns(final double radians) => radians / (2 * math.pi);
  static double radiansToGradians(final double radians) =>
      radians / math.pi * 200;

  static double degreesToRadians(final double degrees) =>
      degrees / 180 * math.pi;
  static double degreesToTurns(final double degrees) => degrees / 360;
  static double degreesToGradians(final double degrees) => degrees / 180 * 200;

  static double turnsToRadians(final double turns) => turns * 2 * math.pi;
  static double turnsToDegrees(final double turns) => turns * 360;
  static double turnsToGradians(final double turns) => turns * 400;

  static double gradiansToRadians(final double gradians) =>
      gradians / 200 * math.pi;
  static double gradiansToDegrees(final double gradians) =>
      gradians / 200 * 180;
  static double gradiansToTurns(final double gradians) => gradians / 400;
}
